Add new "item" StyleBox to the Tree Control

joao-pedro-braz opened the pull request:
## Description
This PR adds a new "item" StyleBox to the Tree Control. Currently, an item will only receive a StyleBox when it's either focused or selected, this PR aims to fix that, and let users specify a StyleBox for the item on it's base state. Besides that, special care has been taken to ensure that the Content Margins are properly handled, acting as padding in both RTL and LTR layouts.

This PR also makes use of this newly added feature to adjust the Editor Theme.
Previously, icons, like the ones in the FileSystemDock or even in the SceneTree, didn't have proper spacing between borders and would therefore appear too close to them, which was specially noticeable with large Corner Radius values, such as 6.

## Comparisons

### Currently (pre-PR)
![image](https://user-images.githubusercontent.com/37230465/222823334-10a53232-a610-4a4c-b170-bdd9202e2ca9.png)
*File System Dock*

![image](https://user-images.githubusercontent.com/37230465/222824464-4f334f6b-8784-4bd0-bf8f-e8ab890359a3.png)
*File System Dock in RTL*

### With PR
![image](https://user-images.githubusercontent.com/37230465/222823632-ffea51ae-fa8c-43f4-aecb-8699bd864a90.png)
*File System Dock*


![image](https://user-images.githubusercontent.com/37230465/222824309-6ad0d32f-c290-4958-841f-efb7877943ed.png)
*File System Dock in RTL*

## Regression Test

Special care was also taken in ensuring that the new StyleBox won't break existing applications.

### Without PR
![image](https://user-images.githubusercontent.com/37230465/222823397-6d481d4a-f5d6-4533-9784-00dda8cf425b.png)
*Application with a default Tree and with it's first element selected*

### With PR
![image](https://user-images.githubusercontent.com/37230465/222823760-bfd791fb-5514-45c2-a35a-4da05e710ef4.png)
*Application with a default Tree and with it's first element selected*

Note: There seems to be a bug in TextLine which causes the text_buf width to be inconsistently reported in RTL layouts. This can be observed as a rapid and frequent change in the text position when resizing the aforementioned FileSystemDock and SceneTree (Only noticeable when the labels are larger than the available space).
This problem was partially addressed by taking into account both the Content Margins
and the icon width (if any) when computing the text_buf's width.

Note 2: This superseeds my [previous PR on the subject](https://github.com/godotengine/godot/pull/74301)

rcorre committed:
Disable local space for blender transforms.

Having local_space enabled when starting a transform changed the
behavior of VIEW space transforms. Now we disable local_space when
starting a blender transform (there was already logic to restore the
setting after the transform ends).

This also hides the gizmo while performing a blender transform,
otherwise the user will see it snap back and forth between the local and
global alignment. I think the transform looks cleaner with the gizmo
hidden anyways.

Fixes #59392.

snailrhymer committed:
Stop pasted child nodes being assigned an owner when previously unowned

Make copy and pasting match duplication's ownership transferral behavior by storing ownership information in the duplicated nodes on the node clipboard, then checking that information when setting owners for pasted nodes.

cooperra committed:
Have the Rename Node action use the targeted Node to determine the current undo/redo context

Formerly, we deduced context implicitly, but this failed and always used the global context instead of the context of the scene containing the Node.

This happened because the first argument to `add_do_method`, the SceneTreeEditor, is a descendant of Node and outside the current game scene's tree (it's part of the editor instead). This led the code in `EditorUndoRedoManager::get_history_id_for_object` to choose global context.

My solution is to explicitly use the renamed Node to deduce our context because it will always be in the current scene in this situation.

Fixes #67276

pkowal1982 committed:
Fix scrolling behaviour with low page value

pfertyk committed:
Fix RichTextLabel crash with out of bound exception

Fixes #68242.

MewPurPur committed:
Improve documentation of int

EricEzaM committed:
Fix cancelling selection (pressing escape) while gizmo editing making uncommitted changes.
* Pressing escape while gizmo editing will discard the changes made during that edit 'session'

marzecdawid committed:
Fix Tree::deselect_all not deselecting root

eskandrej committed:
Fix for deselecting item when select_mode == SELECT_ROW

YeldhamDev committed:
Defer invalidation on FileDialog nodes

Commit:
Create a safe temporary file with is_backup_save_enabled

Sauermann committed:
Fix some ways to create inconsistent Viewport sizes

In the editor, it was possible to set the size of a `SubViewport` even
in cases where a parent `SubViewportContainer` had stretch enabled.

This PR disables editing a `SubViewport.size` while the parent disallows
it and it makes necessary adjustments during `NOTIFICATION_ENTER_TREE`.

bruvzg committed:
Add GDScript `to_wchar_buffer` and `get_string_from_wchar` functions.

MewPurPur committed:
Fix error spam when naming a func at the end of the script

YuriSizov committed:
Reorganize context menu in FileSystem dock to put more used options higher

kilojool committed:
Add missing virtual bind for control::get_tooltip

TheSecondReal0 committed:
Draw Camera2D outlines as 2 point primitives instead of 4

RedMser committed:
Fix clipboard relying on focused window

BastiaanOlij committed:
Move roughness limiter and sort into their own classes

snibk committed:
Redraw 2d viewport when guides are cleared

mihe committed:
Add `pick_ray` parameter to extension binding of `intersect_ray`

joao-pedro-braz committed:
Add new "item" StyleBox to the Tree Control

This PR adds a new "item" StyleBox to the Tree Control.
Currently, an item will only receive a StyleBox when it's either
focused or selected, this PR aims to fix that, and let users
specify a StyleBox for the item on it's base state.
Besides that, special care has been taken to ensure
that the Content Margins are properly handled, acting
as padding in both RTL and LTR layouts.

This PR also makes use of this newly added feature to adjust the
Editor Theme.
Previously, icons, like the ones in the FileSystemDock
or even in the SceneTree, didn't have proper spacing between borders.

Note: There seems to be a bug in TextLine which causes
the text_buf width to be inconsistently reported in RTL layouts.
This can be observed as a rapid and frequent change
in the text position when resizing the aforementioned
FileSystemDock and SystemDock. This problem was partially addressed
by taking into account both the Content Margins
and the icon width (if any) when computing the text_buf's width.

joao-pedro-braz commented:
Do we have any plans on adding Unit Tests for the UI?
Had a look in the tests folder and couldn't find anything related.

joao-pedro-braz committed:
Fix --doctool

(or attempt to, at least)

JohanAR committed:
Fixed incorrect ERROR message when setting new PrimitiveMesh

If calling set_mesh with a PrimitiveMesh that has pending update, the
_mesh_changed function would be called twice. The first time before
set_base had been called, which could lead to an ERROR message about
trying to set an invalid surface override material.

JohanAR committed:
Fix instance uniforms breaking when setting a new mesh.

Fixes #58113

bitsawer committed:
Fix several clang-tidy bool literal conversion warnings

SlugFiller committed:
Fix AABB calculation for meshes using Skeleton2D

hakro committed:
Reload built in script to get its methods on signal connection

stmSi committed:
Fix `Find in Files` Search Results cannot open builtin script

KoBeWi commented:
I think you need to rebase to fix this.

rcorre committed:
Add some missing renames to 3to4 tool.

MultiplayerPeerExtension isn't an exact replacement for
NetworkedMultiplayerCustom, but at least it gets you moving in the right direction.

Engine.editor_hint couldn't be fixed by the renames map, because you have to add a `()` at the end.

ronyeh committed:
Rename mkv_list to mvk_list (MoltenVK path list).

SlugFiller committed:
Fixes a canvas item set to clip children being drawn as black if no children are visible

Rindbee committed:
Fix broken shortcut key input

nongvantinh committed:
Fixes  gutter set width results in receiving only half of the desired size

YuriSizov committed:
Add missing handler for removing font sizes in Themes

YuriSizov committed:
Generate empty textures for theme icons if the SVG module is disabled

RedworkDE committed:
Fix buffer overrun in CPUParticles3D

YuriSizov committed:
Improve logic related to editing audio buses (and prevent crashes)

clayjohn committed:
Propogate errors when creating an OpenGL context fails in X11

YuriSizov committed:
Prevent crashing on startup if project has scripted theme types

Also avoid order of operation conflicts by moving C# binding generation hook to main.cpp

clayjohn committed:
Avoid copying CanvasTexture when updating proxy

timothyqiu committed:
Fix crash when dumping extension API in a non-writable directory

Haydoggo committed:
Fix crash when revealing file in floating FileSystem Dock
When selecting "Show in FileSystem" from the context menu of a resource
in the inspector, the engine would crash if the FileSystem dock was
floating because it was trying to focus the FileSystem tab, but floating
docks don't use Tab Containers. This commit makes the FileSystem dock's
window grab focus instead if it's floating.

bruvzg committed:
[Linux/X11] Add some missing keycodes/scancodes.

timothyqiu committed:
Fix coloring of the renderer options button

rcorre committed:
Don't unset local_coords on starting transform.

This fixes a bug introduced by 806425621cfa3daba3ba9f6b020726490a3152a4, where dragging the gizmo no longer respected local transforms.

I'm not sure why I called set_local_coords_enabled(false) in _compute_edit. Removing this line seems to fix gizmo-dragging local transforms, without breaking anything else.

I also noticed that confirming a transform leaves the gizmo axis lines on the screen. This is fixed by calling update_transform_gizmo after clearing the edit mode/instant flags, so update_transform_gizmo knows not to render any axes.

AThousandShips committed:
Fix Variant hashing for floats

Incorrectly hashed floats as single precision

brunosxs committed:
Fix for AABB methods with wrong description

vnen committed:
Remove outdated note about const in Dictionary and Array docs

YuriSizov committed:
Prevent cache corruption when saving resources in the editor

vonagam committed:
GDScript: Change parser representation of class extends

vonagam committed:
GDScript: Fix missing warning for shadowing of built-in types

Calinou committed:
Tweak command syntax in Vulkan renderer failure message

This quotes the executable name so that copying it always works
(even if the path contains spaces).

The command is also indented from the rest of the text and is
no longer single-quoted, as that can prevent the command from
running if the line is copied in its entirety (with the quotes).

titus125 committed:
Fixed typos in the Transform2D and Transform3D class reference

Haydoggo committed:
Fix errors when closing floating docks

dock->get_index() on line 4463 was not behaving as expected
due to dock having an internal sibling, so now we just get the
index excluding internal nodes.

line 4742 would throw an error if you made multiple docks
floating then redocked the end docks first, but no longer

timothyqiu committed:
Add missing TTRs in tiles editor and array inspector

timothyqiu committed:
Fix various typos in the classref

omar-polo committed:
change some sys.exit() to be fatal

- when not all freetype deps are builtin (or provided externally)
- when attempting to build on an unsupported CPU architecture

akien-mga committed:
Vulkan: Fix VMA build with GCC 13

Fixes #74647.

akien-mga committed:
SCons: Cleanup pulseaudio defines for Linux

No need to define _REENTRANT manually when using the system lib,
it's part of the pkgconfig cflags.
And we were then defining PULSEAUDIO_ENABLED twice.

RedworkDE committed:
Add dedicated setter for `ImageTextureLayered::_images` to fix `create_from_images` being hidden from the C# bindings

mihe committed:
Fix extension bindings for motion collision/result structs

lewiji committed:
Add "filesRoot" path to Android provider paths xml

AThousandShips committed:
Document `editor/naming/scene_name_casing` setting

Moved definitions of editor related project settings to `editor/register_editor_types.cpp` to make documentation work.

vonagam committed:
GDScript: Fix autocomplete inside a block with a type test condition

clayjohn committed:
Add parentheses around arguments when converting xform

bruvzg committed:
[TextServer] Add invalid font scaling check, restrict Linux/BSD system fonts lookup to TrueType/CFF only.

mihe committed:
Add `recovery_as_collision` to extension binding of `_body_test_motion`

BastiaanOlij committed:
Merge duplicate rd_texture functions

yedpodtrzitko committed:
clear filter input in Project Settings when opening Layers

AThousandShips committed:
Exposing more project settings for documentation

justchen1369 committed:
Fix exported type for Menubar start_index

YuriSizov committed:
Make sure Script Debugger is updated with the editor theme

clayjohn committed:
Use linear filtering without mipmaps for ProceduralSkyMaterial and PhysicalSkyMaterial

luzader committed:
OpenXR: Enable access to the Valve Index grip force sensing

KoBeWi committed:
Fix MultiNodeEdit not cleared after deleting nodes

smix8 committed:
Allow negative NavigationAgent2D path debug line_width for thin lines

Allows the line_width for NavigationAgent2D path debug to go negativ for thin line rendering.

clayjohn committed:
Merge pull request #74781 from Malcolmnixon/gles3_android

Fix null-pointer dereference on Android systems when using gl_compatibility renderer

JohnVeness committed:
Corrected capitalisation of macOS

In documentation and comments, ignoring thirdparty code

dsnopek committed:
Documents the "arc-based" Quaternion constructor

pcamp committed:
Fix typo in Window class.

"event as" -> "events will be"

BastiaanOlij committed:
Add a get_system_info method to XRInterface

YuriSizov commented:
@joao-pedro-braz You need to rebase your PR against the master, and then force-push, so that it only contains the changes that you've made.

dalexeev committed:
Fix `InputEventConfigurationDialog` modifies original event

lawnjelly committed:
Improve Vector2 / 3 / 4 normalized() classref.

Mention that the results will be unreliable with initial vector near zero.

rburing committed:
Revert attempted fix of trimesh CCD

fahadshihab committed:
Added drag and drop support for shaderinclude

Dragging and dropping now works for shader include file.

clayjohn committed:
Merge pull request #74831 from JohnVeness/macOS

Corrected naming/capitalisation of macOS

clayjohn committed:
Merge pull request #74847 from pcamp/patch-2

Fix typo in Window class.

smix8 committed:
Fix GridMap free navigation RID error spam

Fixes GridMap free navigation RID error spam.

smix8 committed:
Fix GridMap signal "cell_size_changed" disconnect error

Fixes GridMap signal "cell_size_changed" disconnect error.

BastiaanOlij committed:
XR: When an sRGB target is used, check hardware sRGB conversion

smix8 committed:
Fix agents with disabled avoidance getting added to avoidance simulation

Fixes that agents with disabled avoidance were getting added to avoidance simulation.

Spartan322 committed:
Remove methods in header removed in c7eb814

c7eb814 missed these methods in the RichTextLabel's header
Remove set_override_selected_font_color from rich_text_label.h
Remove is_override_selected_font_color from rich_text_label.h
c7eb814 missed these methods in the TextEdit's header
Remove set_override_selected_font_color from text_edit.h
Remove is_override_selected_font_color from text_edit.h

YuriSizov committed:
Make the request to redraw when clearing guides a part of UndoRedo

AThousandShips committed:
Fix `Array.slice()` rounding for `abs(step) != 1`

paddy-exe committed:
Fix ndc calculation for LinearSceneDepth VS node in GLES3

bruvzg committed:
[macOS] Remove unnecessary debug prints.

Calinou committed:
Add "Bbcode" -> "BBCode" editor capitalization

This property name is used in RichTextLabel.

akien-mga committed:
Merge pull request #74911 from bruvzg/macos_no_dbg_print

[macOS] Remove unnecessary debug prints.

Rindbee committed:
Property list changes are only notified when it did change in Curve

`Object::notify_property_list_changed()` should only be called when
the structure of the property has actually changed.

The structure of the property tree has not changed if the size of the
array has not changed, and the only possibility is that the element
value has changed.

akien-mga committed:
i18n: Sync translations with Weblate

(cherry picked from commit d23922ffebe48f29126c003411495737d07e5a9f)

bruvzg committed:
Remove (or make verbose only) various debug prints.

YuriSizov committed:
Merge pull request #74916 from Calinou/editor-add-bbcode-capitalization

Add "Bbcode" -> "BBCode" editor capitalization

YuriSizov committed:
Merge pull request #74904 from YuriSizov/canvas-item-editor-undraw-redraw

Make the request to redraw when clearing guides a part of UndoRedo

YuriSizov committed:
Merge pull request #74858 from dalexeev/fix-input-event-dialog

Fix `InputEventConfigurationDialog` modifies original event

YuriSizov committed:
Merge pull request #74770 from AThousandShips/min_max

Fix type check for max/min

YuriSizov committed:
Merge pull request #74731 from brettchalupa/patch-1

Fix CharacterBody2D get_slide_collision docs

YuriSizov committed:
Merge pull request #74787 from luzader/openxr-grip-force

OpenXR: Enable access to the Valve Index grip force sensors

YuriSizov committed:
Merge pull request #73144 from chutchinson/bug/openxr-init-alert-dialog

Replaced OpenXR operating system alert dialog with a warning log message

YuriSizov committed:
Merge pull request #74909 from AThousandShips/array_slice

Fix `Array.slice()` rounding for `abs(step) != 1`

YuriSizov committed:
Merge pull request #74927 from Rindbee/notify_property_list_changed_when_it_did

Property list changes are only notified when it did change in Curve

YuriSizov committed:
Merge pull request #74624 from Maran23/converter-theme-overrides

Add conversion for common Theme Overrides

YuriSizov committed:
Merge pull request #74838 from dsnopek/quaternion-constructor-3

Documents the "arc-based" Quaternion constructor

YuriSizov committed:
Merge pull request #74667 from YuriSizov/core-too-many-orphans

Prevent infinite loops when printing orphan nodes

YuriSizov committed:
Merge pull request #74556 from AThousandShips/text_adv_svg_freetype

[TextServer] Make sure Freetype is enabled for ot-svg

YuriSizov committed:
Merge pull request #74660 from and-rad/shader-editor-trim-whitespace

Shader editor trims trailing whitespace if set in editor settings

YuriSizov committed:
Merge pull request #74795 from KoBeWi/MassNodeEdit

Fix MultiNodeEdit not cleared after deleting nodes

YuriSizov committed:
Merge pull request #74735 from KoBeWi/fav_template_place

Remember directory when installing templates file

YuriSizov committed:
Merge pull request #74727 from AThousandShips/project_settings

Exposing more project settings for documentation

YuriSizov committed:
Merge pull request #71307 from eskandrej/master

Fix for deselecting item when select_mode == SELECT_ROW

YuriSizov committed:
Merge pull request #74869 from fahadshihab/master

Added drag and drop support for shader include files in shader editor

YuriSizov committed:
Merge pull request #67590 from cooperra/rename-node-undoredo-context-fix

Have the Rename Node action use the targeted Node for undo/redo context

YuriSizov committed:
Merge pull request #63130 from snailrhymer/copy-duplication-fix

Stop pasted child nodes being assigned an owner when previously unowned

dalexeev committed:
GDScript: Fix false positive `REDUNDANT_AWAIT` warning

clayjohn committed:
Disallow creating a project in the Home or Documents folder

Also don't prompt users to delete the project folder for now.

This is a temporary fix to ensure that users do not delete their entire home folders by mistake

akien-mga committed:
Merge pull request #74964 from clayjohn/pm-file-safety

Disallow creating a project in the Home or Documents folder

timothyqiu committed:
Fix typos and inconsistencies in classref

timothyqiu committed:
Fix error when opening Inspector's dots menu

When the inspector is empty, opening the "extra resource options" menu
produces a error:

> Condition "current_res.is_null()" is true.

Opening the mneu with no current resource is valid use case, so
`ERR_FAIL_COND` should not be used.

smix8 committed:
Fix NavigationAgent3D debug path duplicated points

Fixes duplicated points in NavigationAgent3D debug path.

YuriSizov committed:
Merge pull request #74974 from timothyqiu/red-hot-kebab

Fix error when opening Inspector's dots menu

YuriSizov committed:
Merge pull request #74972 from timothyqiu/classref-fixes

Fix typos and inconsistencies in classref

YuriSizov committed:
Merge pull request #74898 from Spartan322/patch-1

Remove methods in header removed in c7eb814

YuriSizov committed:
Merge pull request #74800 from smix8/navagent2d_thin_line_4.x

Allow negative NavigationAgent2D path debug line_width for thin lines

YuriSizov committed:
Merge pull request #74976 from smix8/navagent_fix_debug_path_duplicates_4.x

Fix NavigationAgent3D debug path duplicated points

YuriSizov committed:
Merge pull request #74892 from BastiaanOlij/fix_hw_srgb_conversion

XR: When an sRGB target is used, check hardware sRGB conversion

YuriSizov committed:
Merge pull request #74910 from paddy-exe/fix-ndc-depth-gles3

Fix ndc calculation for LinearSceneDepth VS node in GLES3

YuriSizov committed:
Merge pull request #73156 from gw2cc/safe_save

Create a safe temporary file with is_backup_save_enabled

YuriSizov committed:
Merge pull request #74626 from KoBeWi/one_line_that_changes_the_fate

Select the newly duplicated file

YuriSizov committed:
Merge pull request #74114 from dalexeev/editor-help-enable-context-menu

Enable `RichTextLabel` context menu if selection is enabled

YuriSizov committed:
Merge pull request #74860 from lawnjelly/normalize_classref

Improve Vector2 / 3 / 4 normalized() classref.

YuriSizov committed:
Merge pull request #73519 from YuriSizov/fsdock-top-down-contextual-order

Reorganize context menu in FileSystem dock to put more used options higher

YuriSizov committed:
Merge pull request #73839 from clayjohn/OUTPUT_IS_SRGB

Clean up OUTPUT_IS_SRGB redefinitions

YuriSizov committed:
Merge pull request #74092 from KoBeWi/zero_reason_to_change

Add early return for changing TileMap properties

YuriSizov committed:
Merge pull request #74149 from bruvzg/get_char

[Font] Implement `get_char_from_glyph_index` function.

YuriSizov committed:
Merge pull request #74066 from m4gr3d/add_vulkan_version_filter_main

Add feature check to require min Vulkan api version 1.0 on Android

kleonc committed:
TileMapEditor Fix preview rendering and transform calculations

YuriSizov committed:
Merge pull request #74564 from YuriSizov/inspector-is-simple-you-see-there-is-a-key-and-we-set-it-and-we-propagate-it-and-it-just-works

Update property keying state without a full Inspector rebuild

akien-mga committed:
Merge pull request #73121 from Scony/add-initial-navi-tests

Add initial navigation tests

KoBeWi committed:
Fix auto-translations in editor

bruvzg committed:
[RTL] Add option to customize list bullet, use U+2022 by default.

YuriSizov committed:
Merge pull request #74861 from rburing/restore_trimesh_ccd

Revert attempted fix of trimesh CCD

mihe committed:
Change documentation of `Shape3D.margin` to reflect current situation

mihe committed:
Open up bounds of `Shape3D.margin` to support other physics engines

MisoMosiSpy committed:
Fix typo in the type-safe examples

Fix type in the type-safe examples of snapped, lerp and sign function.

bruvzg committed:
Fix Unix temp file creations when using is_backup_save_enabled.

adamscott committed:
Set VehiculeWheel3D `suspension_travel` default value to a resonable one

akien-mga committed:
Merge pull request #74684 from clayjohn/export-tex

Delete unused compression formats from .import files when exporting

akien-mga committed:
Merge pull request #75074 from bruvzg/fix_unix_temp_files

Fix Unix temp file creations when using is_backup_save_enabled.

Commit:
Make vulkan level 1 an optional requirement

rcorre committed:
Apply clamp_to_embedder on parent resize and popup.

Fixes #75084.

The clamp_to_embedder setting was added in 8be16e0704873f0c9bab8c10dafe2607a00ff78f,
but was not set on any of the in-editor dialogs.

This patch sets `clamp_to_embedder` on editor dialogs so they cannot be dragged out of the frame.
This also modifies `clamp_to_embedder` so a window is clamped to the bounds of an embedder when
it pops up and when the parent is resized.

Calinou committed:
Improve Color8 documentation

This documents in which situations Color8 is most suited and its
precision limitations.

dsnopek committed:
Add missing documentation for MultiplayerPeerExtension

KoBeWi committed:
Fix get_drag_data not overridable in some Controls

skyace65 committed:
Merge pull request #75076 from MisoMosiSpy/doc_GlobalScope

Fix typo in the type-safe examples

Rindbee committed:
Add instructions for configuring CheckBox as a radio button

YuriSizov committed:
Merge pull request #74931 from bruvzg/no_print

Remove (or make verbose only) various debug prints.

YuriSizov committed:
Merge pull request #75110 from Calinou/doc-color8

Improve Color8 documentation

YuriSizov committed:
Merge pull request #75106 from m4gr3d/set_vulkan_level_optional_main

Make vulkan level 1 an optional requirement

YuriSizov committed:
Merge pull request #74537 from nongvantinh/fix-74458

Fixes gutter set width results in receiving only half of the desired size

YuriSizov committed:
Merge pull request #74442 from rburing/thicc_polygons

Fix concave/convex polygon shape documentation

YuriSizov committed:
Merge pull request #74982 from kleonc/tile-map-editor-fix-preview-transforms

TileMapEditor Fix preview rendering and transform calculations

YuriSizov committed:
Merge pull request #75079 from mihe/shape-margin

Update property/documentation of shape margins

YuriSizov committed:
Merge pull request #75116 from dsnopek/multiplayer-peer-extension-missing-docs

Add missing documentation for MultiplayerPeerExtension

YuriSizov committed:
Merge pull request #74665 from YuriSizov/code-edit-is-a-complete-drag

Prevent passing events from CodeEdit to TextEdit when code completion is active

Commit:
Fix directory access when the running app has the `All files access` permission

YuriSizov committed:
Merge pull request #75080 from adamscott/set-vehiculewheel3d-reasonable-value

Set `VehiculeWheel3D` `suspension_travel` default value to a reasonable one

YuriSizov committed:
Merge pull request #75017 from bruvzg/rlt_list_bullet

[RTL] Add option to customize list bullet, use U+2022 by default.

YuriSizov committed:
Merge pull request #74842 from vonagam/fix-builtin-shadow

GDScript: Fix missing warning for shadowing of built-in types

YuriSizov committed:
Merge pull request #74844 from vonagam/change-class-extends-parsing

GDScript: Change parser representation of class extends

YuriSizov committed:
Merge pull request #74949 from dalexeev/gds-fix-await-warning

GDScript: Fix false positive `REDUNDANT_AWAIT` warning

YuriSizov committed:
Merge pull request #72056 from resistor/master

[GDScript] Perform update-and-assign operations in place when possible.

Pennycook committed:
Add get_loops_left() function to Tween

Implements godotengine/godot-proposals#5141.

Adds a new get_loops_left() function to Tween, allowing developers to
reason about how many times a tweening sequence will repeat and whether
to expect finished or loop_finished as the next signal.

Co-authored-by: Tomek <kobewi4e@gmail.com>

bruvzg committed:
[Windows] Remove meta_mem update on keyup/keydown.

Rindbee committed:
Fix some wrong descriptions in the docs

YuriSizov committed:
Merge pull request #75172 from bruvzg/remove_metamem

[Windows] Remove meta_mem update on keyup/keydown.

YuriSizov committed:
Merge pull request #74890 from smix8/gridmap_cell_size_disconnect_error_4.x

Fix GridMap signal "cell_size_changed" disconnect error

YuriSizov committed:
Merge pull request #74893 from smix8/fix_agents_with_disabled_avoidance_4.x

Fix agents with disabled avoidance getting added to avoidance simulation

YuriSizov committed:
Merge pull request #75122 from KoBeWi/get_drug_data

Fix get_drag_data not overridable in some Controls

YuriSizov committed:
Merge pull request #75134 from Rindbee/configured-as-a-radio-button

Add instructions for configuring CheckBox as a radio button

YuriSizov committed:
Merge pull request #74889 from smix8/gridmap_navrid_error_spam_4.x

Fix GridMap free navigation RID error spam

YuriSizov committed:
Merge pull request #75174 from Rindbee/fix-wrong-descriptions

Fix some wrong descriptions in the docs

YuriSizov committed:
Merge pull request #75146 from m4gr3d/fix_directory_access_with_all_files_access_permission_main

Fix directory access when the running app has the `All files access` permission

YuriSizov committed:
Merge pull request #73800 from YeldhamDev/graph_resize_fix

Fix `GraphNode` resizing when its bottom border is too thin

ator-dev committed:
Fix "Download Project Source" for Web Editor

- Correctly pass safe project name and safe datetime to `vformat`

YuriSizov committed:
Merge pull request #75194 from ator-dev/download-source-fix

Fix "Download Project Source" for Web Editor

AThousandShips committed:
Make `Gradient` resort points on `reverse`

kleonc committed:
CanvasItem Fix ENTER_CANVAS / VISIBILITY_CHANGED notifications order when entering tree

nikitalita committed:
Fix ResourceImporterLayeredTexture::import() high_quality type

AThousandShips committed:
Improve documentation for Area monitor callbacks in `PhysicsServer3D`

AThousandShips committed:
Fix `AnimatedSprite2D` autoplay warning

Changing of autoplay when changing `SpriteFrames` is not done by the
user and warning is unhelpful

rburing committed:
Fix collide_shape return type

Fix PhysicsDirectSpaceState3D::_collide_shape return type.
Also PhysicsDirectSpaceState2D::_collide_shape.

rburing committed:
Port robust signal (dis)connection to ShapeCast2D

Ported from ShapeCast3D.

RedworkDE committed:
C#: Do not print errors about missing references to intentionally ignored members

AThousandShips committed:
Fix "hides member" warnings in `csharp_script.cpp`

Calinou committed:
Document how to use logarithm of base 10 with `log()`

Calinou committed:
Only change floors in GridMap editor when holding Ctrl/Cmd, not Shift

The Shift modifier frequently interfered with freelook's speed modifier,
which led to the floor level being accidentally changed.

YuriSizov committed:
Merge pull request #73897 from TheSecondReal0/camera-line-width

Draw Camera2D outlines as 2 point primitives instead of 4 (consistent with how origin is drawn in 2D editor)

YuriSizov committed:
Merge pull request #75260 from rburing/fix_collide_shape_return

Fix `collide_shape` return type

YuriSizov committed:
Merge pull request #75012 from KoBeWi/english_do_you_speak_it

Fix auto-translations in editor

YuriSizov committed:
Merge pull request #73818 from kilojool/virtual_get_tooltip

Add missing virtual bind for control::get_tooltip

YuriSizov committed:
Merge pull request #75258 from AThousandShips/autoplay

Fix `AnimatedSprite2D` autoplay warning

YuriSizov committed:
Merge pull request #75284 from RedworkDE/net-docs-ref-error

C#: Do not print errors about missing references to intentionally ignored members

YuriSizov committed:
Merge pull request #75235 from AThousandShips/gradient_sort

Make `Gradient` resort points on `reverse`

YuriSizov committed:
Merge pull request #74454 from Pennycook/get_loops_left

Add get_loops_left() function to Tween

YuriSizov committed:
Merge pull request #73225 from bruvzg/wchar_to_from

Add GDScript `to_wchar_buffer` and `get_string_from_wchar` functions.

YuriSizov committed:
Merge pull request #74848 from BastiaanOlij/add_xr_system_info

Add a get_system_info method to XRInterface

YuriSizov committed:
Merge pull request #75251 from AThousandShips/area_doc

Improve documentation for Area monitor callbacks in `PhysicsServer3D`

YuriSizov committed:
Merge pull request #75286 from AThousandShips/csharp_shadow

Fix "hides member" warnings in `csharp_script.cpp`

YuriSizov committed:
Merge pull request #75244 from nikitalita/nikitalita-patch-1

Fix `ResourceImporterLayeredTexture::import()` `high_quality` type

YuriSizov committed:
Merge pull request #71868 from YeldhamDev/filediag_invalidate_deferred

Defer invalidation on FileDialog nodes

KoBeWi committed:
Remove disabled plugins from active plugins

umarcor committed:
[ci/web] update mymindstorm/setup-emsdk to v12

akien-mga committed:
Merge pull request #75339 from umarcor/umarcor/ci/setup-emsdk

[ci/web] update mymindstorm/setup-emsdk to v12

AThousandShips committed:
Fix `NodePath` subname index range documentation

Calinou committed:
Make MarginContainer available with `disable_advanced_gui=yes`

Comparatively more advanced GUI nodes such as FlowContainer and
AspectRatioContainer are already available when building with
`disable_advanced_gui=yes`.

This increases binary size by about 8 KB for binaries compiled with
`disable_advanced_gui=yes` (Linux release export template with LTO enabled).

YuriSizov committed:
Merge pull request #75300 from Calinou/doc-log-log10

Document how to use logarithm of base 10 with `log()`

timothyqiu committed:
Fix height of subresource button in Inspector

timothyqiu committed:
Make Add Scene button the same height as scene tabs

YuriSizov committed:
Merge pull request #75379 from timothyqiu/boundary-break

Fix height of subresource button in Inspector

YuriSizov committed:
Merge pull request #75381 from timothyqiu/add-height

Make Add Scene button the same height as scene tabs

timothyqiu committed:
Auto translate popup menus of MenuButton and OptionButton

Currently, `auto_translate` property of a `MenuButton` or `OptionButton`
won't affect its popup menu. It's okay if you want to auto translate,
but requires extra setup if you don't.

YuriSizov committed:
Merge pull request #75367 from Calinou/disable-advanced-gui-keep-margincontainer

Make MarginContainer available with `disable_advanced_gui=yes`

YuriSizov committed:
Merge pull request #75331 from KoBeWi/remove_removed_plugins

Remove disabled plugins from active plugins

timothyqiu committed:
More i18n improvements

* Make placeholder in editor layout dialog translatable.
* Make messages in scene import settings dialog translatable.
* Mark theme override property categories for translation.

YuriSizov committed:
Merge pull request #71405 from marzecdawid/deselect-root-in-deselect_all

Fix `Tree::deselect_all` not deselecting root

YuriSizov committed:
Merge pull request #75350 from AThousandShips/nodepath_doc_fix

Fix `NodePath` subname index range documentation

YuriSizov committed:
Merge pull request #75238 from kleonc/canvas-item-fix-notifications-order-on-enter-tree

`CanvasItem` Fix `ENTER_CANVAS` / `VISIBILITY_CHANGED` notifications order when entering tree

YuriSizov committed:
Merge pull request #75266 from rburing/shapecast2d_disconnect

Port robust signal (dis)connection to `ShapeCast2D`

timothyqiu committed:
Fix typos and inconsistencies in classref

* Typo fixes.
* Navigation agent's `velocity_computed` signal is not emitted by
  `set_velocity`, but at the end of that frame if `set_velocity` is
  called.
* Simplify link in navigation agent's `avoidance_enabled` description.
* Unify `navigation_layers` description across `NavigationAgent{2,3}D`.
* Unify `is_normalized()` description across `Vector2`, `Vector3`, and
  `Vector4`

YuriSizov committed:
Merge pull request #75384 from timothyqiu/cultural-heritage

Auto translate popup menus of MenuButton and OptionButton

YuriSizov committed:
Merge pull request #75304 from Calinou/gridmap-editor-change-floor-ctrl-only

Only change floors in GridMap editor when holding Ctrl/Cmd, not Shift

YuriSizov committed:
Merge pull request #72361 from the-brickster/master

Incorporating the availability of screen and depth textures for the GLES3 backend

YuriSizov committed:
Merge pull request #70046 from MewPurPur/improve-number-docs

Improve documentation of int

YuriSizov committed:
Merge pull request #75385 from timothyqiu/i18n

More i18n improvements

YuriSizov committed:
Merge pull request #75389 from timothyqiu/typo-fixes

Fix typos and inconsistencies in classref

newobj committed:
Update GPUParticles2D/3D speed scale on ENTER_TREE

Fix for https://github.com/godotengine/godot/issues/75218

Pause notifications are not sent when a node is added as a child. So GPUParticles2D should also obey its can_process status on ENTER_TREE, not just PAUSED/UNPAUSED.

YuriSizov committed:
Merge pull request #75141 from rcorre/rrc/dialog-pos

YuriSizov committed:
Merge pull request #75398 from newobj/issue-75218

Update GPUParticles2D/3D speed scale on ENTER_TREE

joao-pedro-braz committed:
Use MODELVIEW_MATRIX for vertex transform when double precision is enabled

This commit updates the double precision vertex transform code from
using the MODEL_MATRIX to now use the MODELVIEW_MATRIX instead.

This can be made possible by transforming the MODELVIEW_MATRIX
back into model space (ie, same space as the MODEL_MATRIX) and then using it as if it were the MODEL_MATRIX.

With this in place we now properly handle VERTEX transformations that
a Material Shader might do, such as billboard-ing.

joao-pedro-braz committed:
Merge remote-tracking branch 'origin/master'

# Conflicts:
#	doc/classes/Tween.xml
#	editor/project_converter_3_to_4.cpp
#	editor/register_editor_types.cpp
#	platform/macos/display_server_macos.mm
